File: examples/hosts/python/src/openwop_host/webhooks.py
"""Webhook subscriptions + signed delivery (stdlib-only port).

Mirrors `examples/hosts/sqlite/src/webhooks.ts`. Implements the webhook
surface from `spec/v1/webhooks.md`:

    - POST   /v1/webhooks          register
    - DELETE /v1/webhooks/{id}     unregister
    - Best-effort HTTP POST delivery on every run event
    - HMAC-SHA256 signing over `{timestamp}.{rawBody}` with headers
        X-openwop-Signature
        X-openwop-Signature-Timestamp
        X-openwop-Signature-Algorithm: v1
        X-openwop-Subscription-Id

Reference-only properties:

    - In-process delivery (no queue); errors swallowed.
    - No retry / backoff / circuit breaker.
    - Per-subscription secret generated when caller omits one; returned
      exactly once on register.
    - SSRF guard rejects loopback / RFC1918 / link-local / unique-local
      / `*.local|*.internal|*.cluster|localhost`. Bypass with
      `OPENWOP_WEBHOOK_ALLOW_PRIVATE=true` for local-receiver tests.

All stdlib: `hmac`, `hashlib`, `secrets`, `ipaddress`, `urllib.request`,
`threading`.
"""

from __future__ import annotations

import hashlib
import hmac
import ipaddress
import json
import os
import secrets
import threading
import time
import urllib.error
import urllib.request
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any
from urllib.parse import urlparse


class WebhookUrlRejected(Exception):
    """SSRF-guard rejection of a webhook URL."""

    def __init__(self, reason: str) -> None:
        super().__init__(reason)
        self.reason = reason


def _url_allowed(raw_url: str) -> tuple[bool, str]:
    # URL shape validation runs UNCONDITIONALLY — `OPENWOP_WEBHOOK_ALLOW_PRIVATE`
    # only relaxes the private-IP / internal-hostname checks for tests
    # against a local receiver. Malformed input is always rejected so a
    # daemon delivery thread never sees a non-parseable URL.
    try:
        parsed = urlparse(raw_url)
    except ValueError:
        return False, "webhook url is not a parseable URL"
    if parsed.scheme not in ("http", "https"):
        return False, f'webhook url protocol "{parsed.scheme}" is not http(s)'
    host = (parsed.hostname or "").lower()
    if not host:
        return False, "webhook url is missing a hostname"

    if os.environ.get("OPENWOP_WEBHOOK_ALLOW_PRIVATE") == "true":
        # Bypass the SSRF heuristics for local-receiver test setups
        # (webhook-signed-delivery.test.ts boots an HTTP receiver on
        # 127.0.0.1). URL shape is already validated above.
        return True, ""

    if host == "localhost" or host.endswith(".localhost"):
        return False, "webhook url points at localhost (SSRF guard)"
    if host.endswith(".local") or host.endswith(".internal") or host.endswith(".cluster"):
        return False, f'webhook url hostname "{host}" looks internal (SSRF guard)'

    try:
        ip = ipaddress.ip_address(host)
    except ValueError:
        # Not an IP literal — we do NOT resolve DNS here (documented
        # limitation; mirrors the TS reference host).
        return True, ""

    if ip.is_loopback:
        return False, "loopback address (SSRF guard)"
    if ip.is_private:
        return False, "RFC1918 / unique-local address (SSRF guard)"
    if ip.is_link_local:
        return False, "link-local address (AWS metadata range) (SSRF guard)"
    if ip.is_unspecified:
        return False, "0.0.0.0/:: (SSRF guard)"
    return True, ""


# The single tenant scope this reference host's runs execute under.
# webhooks.md §Endpoints + RFC 0093 §A.3: the tenant established at
# registration time scopes both who may manage the subscription and
# which run events it receives.
DEFAULT_TENANT_ID = "tenant:default"


@dataclass
class WebhookSubscription:
    subscription_id: str
    url: str
    secret: str
    event_types: list[str]  # empty = subscribe to all events
    created_at: str
    # Tenant scope established at registration (RFC 0093 §A.3).
    tenant_id: str = DEFAULT_TENANT_ID

    def to_public_dict(self) -> dict[str, Any]:
        """Shape returned on POST /v1/webhooks (secret included exactly once)."""
        return {
            "subscriptionId": self.subscription_id,
            # webhooks.md §Register response — canonical id field. Kept
            # alongside the host's historical `subscriptionId` alias.
            "webhookId": self.subscription_id,
            "url": self.url,
            "secret": self.secret,
            "eventTypes": list(self.event_types),
            "createdAt": self.created_at,
            "tenantId": self.tenant_id,
        }


@dataclass
class WebhookRegistry:
    """Thread-safe in-memory webhook-subscription store."""

    _lock: threading.Lock = field(default_factory=threading.Lock)
    _subs: dict[str, WebhookSubscription] = field(default_factory=dict)

    def register(
        self,
        url: str,
        *,
        secret: str | None = None,
        event_types: list[str] | None = None,
        tenant_id: str = DEFAULT_TENANT_ID,
    ) -> WebhookSubscription:
        ok, reason = _url_allowed(url)
        if not ok:
            raise WebhookUrlRejected(reason)
        sub = WebhookSubscription(
            subscription_id=f"wh-{uuid.uuid4()}",
            url=url,
            secret=secret or secrets.token_urlsafe(32),
            event_types=list(event_types or []),
            created_at=datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
            tenant_id=tenant_id,
        )
        with self._lock:
            self._subs[sub.subscription_id] = sub
        return sub

    def unregister(self, subscription_id: str, tenant_id: str = DEFAULT_TENANT_ID) -> bool:
        """Remove a subscription within a tenant scope.

        Scoped to ``tenant_id`` so a subscription held under one tenant
        can never be removed through another tenant's scope (RFC 0093
        §A.3); a non-member caller is rejected with 403 by the HTTP
        handler BEFORE this runs, so no existence information leaks.
        """
        with self._lock:
            sub = self._subs.get(subscription_id)
            if sub is None or sub.tenant_id != tenant_id:
                return False
            del self._subs[subscription_id]
            return True

    def matches(self, event_type: str, tenant_id: str = DEFAULT_TENANT_ID) -> list[WebhookSubscription]:
        """Subscriptions for `event_type` **within `tenant_id`** (RFC 0093 §A.3:
        a subscription receives only events from runs within its tenant scope)."""
        with self._lock:
            snapshot = list(self._subs.values())
        return [
            s
            for s in snapshot
            if s.tenant_id == tenant_id and (not s.event_types or event_type in s.event_types)
        ]


def sign_payload(secret: str, timestamp: str, raw_body: str) -> str:
    """HMAC-SHA256 hex digest of `{timestamp}.{rawBody}` per webhooks.md.

    The same recipe the TypeScript host emits and the conformance suite
    verifies — making cross-host signature parity mechanical.
    """
    mac = hmac.new(
        secret.encode("utf-8"),
        f"{timestamp}.{raw_body}".encode("utf-8"),
        hashlib.sha256,
    )
    return mac.hexdigest()


def _redact_for_fan_out(event: dict[str, Any]) -> dict[str, Any]:
    """Strip the open-ended `data` field per debug-bundle.md redaction policy.

    Receivers get the *envelope* (type, runId, seq, timestamp, nodeId)
    but NOT the open-ended `data` field which may carry interrupt
    configs, approval payloads, or BYOK credentialRefs. To fetch the
    full payload, receivers MUST use an authenticated GET /v1/runs/{id}
    or the debug-bundle endpoint.
    """
    return {k: v for k, v in event.items() if k != "payload"}


def deliver(sub: WebhookSubscription, payload: dict[str, Any], *, timeout_s: float = 5.0) -> None:
    """Fire-and-forget HTTP POST. All errors are swallowed.

    The caller has no way to surface delivery failures to the workflow
    or to the registering tenant — best-effort delivery per
    `webhooks.md` §"Delivery semantics". Production hosts would queue
    retries with backoff + a circuit breaker; the reference impl logs
    nothing and moves on.
    """
    body = json.dumps(payload)
    timestamp = str(int(time.time()))
    signature = sign_payload(sub.secret, timestamp, body)
    try:
        request = urllib.request.Request(
            sub.url,
            data=body.encode("utf-8"),
            method="POST",
            headers={
                "Content-Type": "application/json",
                "X-openwop-Signature": signature,
                "X-openwop-Signature-Timestamp": timestamp,
                "X-openwop-Signature-Algorithm": "v1",
                "X-openwop-Subscription-Id": sub.subscription_id,
            },
        )
        with urllib.request.urlopen(request, timeout=timeout_s) as resp:
            resp.read()  # drain
    except (urllib.error.URLError, TimeoutError, OSError, ValueError):
        # ValueError covers urllib's "unknown url type" raised when a
        # subscription was somehow registered with a malformed URL
        # (defense-in-depth — _url_allowed already rejects these on
        # register).
        return


def fan_out_event(
    registry: WebhookRegistry,
    event: dict[str, Any],
    tenant_id: str = DEFAULT_TENANT_ID,
) -> None:
    """Fan out a single event to every matching subscriber **within the
    event's tenant scope** in background threads. This reference host
    executes every run under ``DEFAULT_TENANT_ID``, so deliveries are
    filtered to subscriptions registered under that scope (RFC 0093 §A.3).
    """
    subs = registry.matches(event.get("type", ""), tenant_id)
    if not subs:
        return
    safe = _redact_for_fan_out(event)
    for sub in subs:
        # Daemon thread so process shutdown doesn't block on outstanding
        # delivery attempts.
        t = threading.Thread(
            target=deliver,
            args=(sub, safe),
            daemon=True,
            name=f"webhook-deliver-{sub.subscription_id}",
        )
        t.start()

File: examples/hosts/postgres/src/webhooks.ts
/**
 * Webhook subscriptions + signed-delivery for the Postgres reference host.
 *
 * Async-pg port of `examples/hosts/sqlite/src/webhooks.ts`. Implements
 * the webhook surface from `spec/v1/webhooks.md`:
 *   - `POST /v1/webhooks` register
 *   - `DELETE /v1/webhooks/{id}` unregister
 *   - HMAC-SHA256 over `{timestamp}.{rawBody}` with `X-openwop-Signature*`
 *     and `X-openwop-Signature-Algorithm: v1` headers.
 *   - Best-effort fire-and-forget delivery; receivers fetch authenticated
 *     run snapshot or debug-bundle for the full payload.
 *   - SSRF guard rejects RFC1918 / link-local / loopback / .local /
 *     .internal / .cluster hosts (bypass via OPENWOP_WEBHOOK_ALLOW_PRIVATE).
 *
 * Postgres-specific: JSONB column for `event_types` (pg-types auto-parses
 * on read, so the SQLite port's JSON.parse is dropped). Every fn that
 * touches the DB is async-pg through the Querier interface.
 *
 * @see spec/v1/webhooks.md
 * @see examples/hosts/sqlite/src/webhooks.ts — the source of the port
 */

import { createHmac, randomBytes, randomUUID } from 'node:crypto';
import { request as httpRequest } from 'node:http';
import { request as httpsRequest } from 'node:https';
import { isIP } from 'node:net';
import type { Querier } from './db.js';

/**
 * SSRF guard for webhook destination URLs. Rejects URLs whose hostname
 * points at loopback / RFC1918 / link-local / unique-local addresses or
 * the common "internal-only" hostnames. Bypass: set
 * OPENWOP_WEBHOOK_ALLOW_PRIVATE=true (for the host's own conformance
 * scenarios where the test receiver runs on localhost).
 *
 * No DNS resolution — a production host would resolve and re-check, and
 * re-resolve at delivery time to defend against DNS rebinding. Reference-
 * impl limitation.
 */
function urlAllowed(rawUrl: string): { ok: true } | { ok: false; reason: string } {
  if (process.env.OPENWOP_WEBHOOK_ALLOW_PRIVATE === 'true') return { ok: true };

  let url: URL;
  try {
    url = new URL(rawUrl);
  } catch {
    return { ok: false, reason: 'webhook url is not a parseable URL' };
  }
  if (url.protocol !== 'http:' && url.protocol !== 'https:') {
    return { ok: false, reason: `webhook url protocol "${url.protocol}" is not http(s)` };
  }

  const host = url.hostname.toLowerCase();
  const bare = host.startsWith('[') && host.endsWith(']') ? host.slice(1, -1) : host;

  if (bare === 'localhost' || bare.endsWith('.localhost')) {
    return { ok: false, reason: 'webhook url points at localhost (SSRF guard)' };
  }
  if (bare.endsWith('.local') || bare.endsWith('.internal') || bare.endsWith('.cluster')) {
    return { ok: false, reason: `webhook url hostname "${bare}" looks internal (SSRF guard)` };
  }

  const ipVersion = isIP(bare);
  if (ipVersion === 4) {
    const [a, b] = bare.split('.').map(Number) as [number, number];
    if (a === 127) return { ok: false, reason: '127.0.0.0/8 (loopback) (SSRF guard)' };
    if (a === 10) return { ok: false, reason: '10.0.0.0/8 (RFC1918) (SSRF guard)' };
    if (a === 172 && b >= 16 && b <= 31) {
      return { ok: false, reason: '172.16.0.0/12 (RFC1918) (SSRF guard)' };
    }
    if (a === 192 && b === 168) {
      return { ok: false, reason: '192.168.0.0/16 (RFC1918) (SSRF guard)' };
    }
    if (a === 169 && b === 254) {
      return { ok: false, reason: '169.254.0.0/16 (link-local; AWS metadata) (SSRF guard)' };
    }
    if (a === 0) return { ok: false, reason: '0.0.0.0/8 (SSRF guard)' };
  }
  if (ipVersion === 6) {
    const lower = bare;
    if (lower === '::1' || lower === '::ffff:127.0.0.1') {
      return { ok: false, reason: 'IPv6 loopback (SSRF guard)' };
    }
    if (lower.startsWith('fc') || lower.startsWith('fd')) {
      return { ok: false, reason: 'fc00::/7 (unique local) (SSRF guard)' };
    }
    if (
      lower.startsWith('fe8') ||
      lower.startsWith('fe9') ||
      lower.startsWith('fea') ||
      lower.startsWith('feb')
    ) {
      return { ok: false, reason: 'fe80::/10 (link-local) (SSRF guard)' };
    }
  }

  return { ok: true };
}

/**
 * Strip payload-bearing fields from a run event before fan-out. Receivers
 * get the envelope (type, runId, seq, timestamp, nodeId) but NOT the
 * open-ended `data` field — that may carry interrupt configs, approval
 * payloads, or future BYOK credentialRefs. Receivers fetch full payloads
 * via authenticated `GET /v1/runs/{id}` or `/debug-bundle`.
 */
function redactForFanOut(event: { type: string; [k: string]: unknown }): Record<string, unknown> {
  const { data, ...rest } = event;
  void data;
  return rest;
}

/**
 * The single tenant scope this reference host's runs execute under.
 * webhooks.md §Endpoints + RFC 0093 §A.3: the tenant established at
 * registration time scopes both who may manage the subscription and
 * which run events it receives.
 */
export const DEFAULT_TENANT_ID = 'tenant:default';

export interface WebhookSubscription {
  readonly subscriptionId: string;
  readonly url: string;
  readonly secret: string;
  readonly eventTypes: ReadonlyArray<string>;
  readonly createdAt: string;
  /** Tenant scope established at registration (RFC 0093 §A.3). */
  readonly tenantId: string;
}

interface SubscriptionRow {
  subscription_id: string;
  url: string;
  secret: string;
  /** JSONB → already-parsed array of strings (or [] default). */
  event_types: string[] | null;
  created_at: string;
  tenant_id: string;
}

export async function setupWebhookSchema(q: Querier): Promise<void> {
  await q.query(`
    CREATE TABLE IF NOT EXISTS webhook_subscriptions (
      subscription_id TEXT PRIMARY KEY,
      url TEXT NOT NULL,
      secret TEXT NOT NULL,
      event_types JSONB NOT NULL DEFAULT '[]'::JSONB,
      created_at TEXT NOT NULL,
      tenant_id TEXT NOT NULL DEFAULT '${DEFAULT_TENANT_ID}'
    );
  `);
  // Migration for databases created before the RFC 0093 §A.3 tenant column.
  await q.query(
    `ALTER TABLE webhook_subscriptions ADD COLUMN IF NOT EXISTS tenant_id TEXT NOT NULL DEFAULT '${DEFAULT_TENANT_ID}'`,
  );
}

export interface RegisterInput {
  readonly url: string;
  readonly secret?: string;
  readonly eventTypes?: ReadonlyArray<string>;
  /** Tenant scope the subscription lives under. The HTTP handler MUST
   *  have already verified the caller's membership (webhooks.md
   *  §Register: "Caller MUST be a member"). */
  readonly tenantId: string;
}

export class WebhookUrlRejected extends Error {
  constructor(public readonly reason: string) {
    super(reason);
    this.name = 'WebhookUrlRejected';
  }
}

export async function registerWebhook(
  q: Querier,
  input: RegisterInput,
): Promise<WebhookSubscription> {
  const guard = urlAllowed(input.url);
  if (!guard.ok) throw new WebhookUrlRejected(guard.reason);

  const subscriptionId = `wh-${randomUUID()}`;
  const secret = input.secret ?? randomBytes(32).toString('base64url');
  const eventTypes = Array.isArray(input.eventTypes) ? input.eventTypes : [];
  const createdAt = new Date().toISOString();
  await q.query(
    `INSERT INTO webhook_subscriptions (subscription_id, url, secret, event_types, created_at, tenant_id)
     VALUES ($1, $2, $3, $4, $5, $6)`,
    [subscriptionId, input.url, secret, JSON.stringify(eventTypes), createdAt, input.tenantId],
  );
  return { subscriptionId, url: input.url, secret, eventTypes, createdAt, tenantId: input.tenantId };
}

/**
 * Remove a subscription within a tenant scope. The delete is scoped to
 * `tenantId` so a subscription held under one tenant can never be
 * removed through another tenant's scope (RFC 0093 §A.3); a non-member
 * caller is rejected with 403 by the HTTP handler BEFORE this runs, so
 * no existence information leaks across tenants.
 */
export async function unregisterWebhook(
  q: Querier,
  subscriptionId: string,
  tenantId: string,
): Promise<boolean> {
  const res = await q.query(
    'DELETE FROM webhook_subscriptions WHERE subscription_id = $1 AND tenant_id = $2',
    [subscriptionId, tenantId],
  );
  return (res.rowCount ?? 0) > 0;
}

/**
 * Load subscriptions matching `eventType`. Reference-impl design: read
 * every row + filter in JS. Production hosts with many subscribers
 * SHOULD denormalize to a `webhook_event_subscriptions` join table so
 * the SQL plan can index-scan by event_type.
 */
async function loadSubscriptionsForEvent(
  q: Querier,
  eventType: string,
  tenantId: string,
): Promise<WebhookSubscription[]> {
  const res = await q.query<SubscriptionRow>(
    'SELECT * FROM webhook_subscriptions WHERE tenant_id = $1',
    [tenantId],
  );
  return res.rows
    .map((r) => ({
      subscriptionId: r.subscription_id,
      url: r.url,
      secret: r.secret,
      eventTypes: (r.event_types ?? []) as string[],
      createdAt: r.created_at,
      tenantId: r.tenant_id,
    }))
    .filter((s) => s.eventTypes.length === 0 || s.eventTypes.includes(eventType));
}

/**
 * Sign a payload per webhooks.md §"Signature scheme". Hex-encoded
 * HMAC-SHA256 of `${timestamp}.${rawBody}`.
 */
export function signPayload(secret: string, timestamp: string, rawBody: string): string {
  return createHmac('sha256', secret)
    .update(`${timestamp}.${rawBody}`, 'utf8')
    .digest('hex');
}

/**
 * Best-effort POST to a single subscriber. Errors swallowed. Returns a
 * promise that resolves when the request completes (success or failure).
 */
function deliver(sub: WebhookSubscription, payload: unknown): Promise<void> {
  return new Promise((resolve) => {
    let url: URL;
    try {
      url = new URL(sub.url);
    } catch {
      resolve();
      return;
    }
    const body = JSON.stringify(payload);
    const timestamp = Math.floor(Date.now() / 1000).toString();
    const signature = signPayload(sub.secret, timestamp, body);

    const opts = {
      method: 'POST',
      hostname: url.hostname,
      port: url.port || (url.protocol === 'https:' ? 443 : 80),
      path: url.pathname + url.search,
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(body),
        'X-openwop-Signature': signature,
        'X-openwop-Signature-Timestamp': timestamp,
        'X-openwop-Signature-Algorithm': 'v1',
        'X-openwop-Subscription-Id': sub.subscriptionId,
      },
    };
    const reqFn = url.protocol === 'https:' ? httpsRequest : httpRequest;
    const req = reqFn(opts, (res) => {
      res.resume();
      res.on('end', () => resolve());
    });
    req.on('error', () => resolve());
    req.setTimeout(5000, () => {
      req.destroy();
      resolve();
    });
    req.write(body);
    req.end();
  });
}

/**
 * Fan out a single event to every matching subscriber **within the
 * event's tenant scope**. Fire-and-forget; caller does not await
 * delivery. RFC 0093 §A.3: a subscription MUST receive only events from
 * runs within its tenant scope; this reference host executes every run
 * under `DEFAULT_TENANT_ID`, so deliveries are filtered to
 * subscriptions registered under that scope.
 */
export async function fanOutEvent(
  q: Querier,
  event: { type: string; [k: string]: unknown },
  tenantId: string = DEFAULT_TENANT_ID,
): Promise<void> {
  const subs = await loadSubscriptionsForEvent(q, event.type, tenantId);
  if (subs.length === 0) return;
  const safe = redactForFanOut(event);
  for (const sub of subs) {
    void deliver(sub, safe).catch(() => undefined);
  }
}
